modules/profiling/src/main/java/org/apache/ignite/internal/profiling/util/ProfilingDeserializer.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.profiling.util;
+
+import java.nio.ByteBuffer;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.ignite.internal.processors.cache.query.GridCacheQueryType;
+import org.apache.ignite.internal.profiling.IgniteProfiling;
+import org.apache.ignite.internal.profiling.IgniteProfiling.CacheOperationType;
+import org.apache.ignite.internal.profiling.LogFileProfiling;
+import org.apache.ignite.internal.profiling.LogFileProfiling.OperationType;
+import org.apache.ignite.internal.util.GridIntList;
+import org.apache.ignite.lang.IgniteUuid;
+
+import static org.apache.ignite.internal.profiling.LogFileProfiling.readIgniteUuid;
+import static org.apache.ignite.internal.profiling.LogFileProfiling.readUuid;
+
+/**
+ * Profiling operations deserializer.
+ *
+ * @see LogFileProfiling
+ */
+public class ProfilingDeserializer implements AutoCloseable {
+    /** Cached strings by id. */
+    private final ConcurrentHashMap<Short, String> stringById = new ConcurrentHashMap<>();
+
+    /** Handlers to process deserialized operation. */
+    private final IgniteProfiling[] handlers;
+
+    /** @param handlers Handlers to process deserialized operation. */
+    public ProfilingDeserializer(IgniteProfiling... handlers) {
+        this.handlers = handlers;
+    }
+
+    /**
+     * Tries to deserialize profiling operation from buffer.
+     *
+     * @param buf Buffer.
+     * @return {@code True} if operation parsed. {@code False} if not enough bytes.
+     */
+    public boolean deserialize(ByteBuffer buf) {
+        int pos = buf.position();
+
+        if (buf.remaining() < 1)
+            return false;
+
+        byte opTypeByte = buf.get();
+
+        OperationType opType = OperationType.fromOrdinal(opTypeByte);
+
+        switch (opType) {
+            case CACHE_OPERATION: {
+                if (buf.remaining() < 1 + 4 + 8 + 8)
```

Comment:
Such calculations are present at least in LogFileProfiling too, need to extract them to a single place to avoid errors.

Reply:
I put the writer and iterator in one place. See `FileProfiling` and `FileProfilingWalker`. Add tests. See `AbstractProfilingTest` implementations.